### PR TITLE
tt: add compatibility of pack command with the cartridge cli pack

### LINKED ISSUE #504: Add the compatibility of the `pack` module with the cartridge cli `pack`
Since we want to maintain backward compatibility with the cartridge cli (within reason), it would be nice to make the result of the `pack` module as similar as possible to that of the same cartridge cli module.
The main difference is that `tt pack` packs the environment and `cartridge cli` packs the application. So I propose to make a compatibility layer under the `--cartridge-compat` option.

Right now I know the following differences:
- `tt` takes on board all `.git' stuff (to be resolved https://github.com/tarantool/tt/issues/284)
- naming differences (`myapp1-1.1.1-0-g76af547.x86_64.tar.gz` vs `myapp1-1.1.1.0.x86_64.tar.gz`)
1. `tt` uses the environment name instead of the application name in the case of `cartridge cli`
2. `tt` adds a commit hash to the version (`1.1.1-0-g76af547` vs `1.1.1.0`)
3. version formats have differences (`1.1.1-0` vs `1.1.1.0`)
- Differences in directory structure
1. `tt` packs binaries (`tt` and `tarantool`) into `bin` directory instead of `root` in case of `cartridge cli` (`tarantool` and `tarantoolctl` (in case of SDK).
2. `cartridge cli` packs `tt` and `tarantool` from `PATH` in case of SDK. `tt` gets binaries from env if it has been installed over `tt install`.
3. in case of `tt` `VERSION` ans `VERSION.lua` files are absent

### PR DESCRIPTION
**Pull request by DerekBum**
New flag `--cartridge-compat` added for pack command, that fixes some differences between `tt pack` and `cartridge-cli pack`.
It solves naming differences (of resulting packages and applications in them). Also `VERSION` and `VERSION.lua` files will be present in resulting bundle.

Closes #504

### CODE REVIEW

File: cli/pack/common.go [hunk 1]
```diff
@@ -448,6 +525,16 @@ func getPackageName(packCtx *PackCtx, opts *config.CliOpts, suffix string,
 	var packageName string
 
 	if packCtx.FileName != "" {
+		if packCtx.CartridgeCompat {
+			// We need more info for generating VERSION and VERSION.lua files
```
Comment (psergee) [inline cli/pack/common.go:529]: Which info?

**Review by psergee — COMMENTED** [review]
Thanks for the patch.
Please, add integration tests for pack with new option. Scenarios to check:
- Pack single app. Should work.
- Pack multiple apps in instances enabled, but only one specified in --app-list. Should work.
- Pack multiple apps. Shouldn't work. Cannot pack multiple apps in compat mode.
- Pack app from current directory (`instances_enabled` is set to '.')

File: cli/cmd/pack.go [hunk 2]
```diff
@@ -51,6 +51,8 @@ The supported types are: tgz, deb, rpm`,
 		packCtx.WithoutBinaries, "Don't include tarantool and tt binaries to the result package")
 	packCmd.Flags().BoolVar(&packCtx.WithBinaries, "with-binaries", packCtx.WithoutBinaries,
 		"Include tarantool and tt binaries to the result package")
+	packCmd.Flags().BoolVar(&packCtx.CartridgeCompat, "cartridge-compat", false,
```
Comment (psergee) [inline cli/cmd/pack.go:54]: Do we check somewhere this is valid only for tgz pack type?

File: cli/pack/common.go [hunk 3]
```diff
@@ -177,18 +204,23 @@ func createPackageStructure(destPath string) error {
...
-	appPath, err := filepath.EvalSymlinks(appPath)
+func copyAppSrc(appPath string, appName string, packagePath string) error {
+	previousPath := appPath
+	appPath, err := filepath.EvalSymlinks(previousPath)
 	if err != nil {
 		return err
 	}
 
+	if previousPath != appPath {
+		appName = filepath.Base(appPath)
```
[the remark this answers is not in the document]
Comment (DerekBum) [reply cli/pack/common.go:215]: Yes, and there is an unit test about that. Suppose we have applications `app1.lua` and `app3.lua`, and `app3.lua` symlinked to `app1.lua`. 
So when we copying those apps, we want to see only one application in the bundle: `app1.lua`. But for `app3.lua` we will call `copyAppSrc` function with `appName == app3.lua` and copy `app3.lua`, so there will be two applications instead. We can fix this by updating `appName`, if there is symlink found.

[the remark this answers is not in the document]
Comment (psergee) [reply cli/pack/common.go:215]: What will archive name in this case?

[the remark this answers is not in the document]
Comment (DerekBum) [reply cli/pack/common.go:215]: Well, it depends on other flags. If we used `--filename` flag, we also specified the name of the archive.
If we specified the package name via `--name` flag, archive will be named as `<name>-<version>.<arch>.tar.gz`.
Else, folder name of the environment will be used as a name in `<name>-<version>.<arch>.tar.gz`.

`copyAppSrc` function has no affect on archive name

**Comment by psergee** [comment]
- After unpacking with `--strip-components=1`, looks like `tt.yaml` disappears.
- The structure of the archive still differs from `cartridge`. In cartridge archive there is only one directory in root - `<app_name>`. `tt pack` must create the same structure. `tt init` in cartridge application directory will create the desired structure.

File: test/integration/pack/test_pack.py [hunk 4]
```diff
@@ -156,6 +156,49 @@ def prepare_tgz_test_cases(tt_cmd) -> list:
             ],
             "artifacts_in_separated_dir": False,
         },
+        {
+            "bundle_src": "bundle8",
+            "cmd": tt_cmd,
+            "pack_type": "tgz",
+            "args": ["--cartridge-compat"],
+            "res_file": "app_name-0.1.0.0." + get_arch() + ".tar.gz",
+            "check_exist": [
```
Comment (psergee) [inline test/integration/pack/test_pack.py:165]: Need to check that no other files/dirs exist in archive root. Only app name dir.

**Comment by psergee** [comment]
Manual test:
```
$ tar xf ../app1-0.1.0.0.x86_64.tar.gz --strip-components=1

$ tt start
   • Starting an instance [tmp:s2-master]...
   • Starting an instance [tmp:s2-replica]...
   • Starting an instance [tmp:stateboard]...
   • Starting an instance [tmp:router]...
   • Starting an instance [tmp:s1-master]...
   • Starting an instance [tmp:s1-replica]...

$ tt status
INSTANCE           STATUS          PID
tmp:stateboard     RUNNING         11275
tmp:router         NOT RUNNING     
tmp:s1-master      NOT RUNNING     
tmp:s1-replica     NOT RUNNING     
tmp:s2-master      NOT RUNNING     
tmp:s2-replica     NOT RUNNING
```
Suppose, this is due to app name detection difference in `tt` and `cartridge`. Need to investigate the way cartridge finds out the application name. From the rockspec file probably...

File: cli/pack/common.go [hunk 5]
```diff
@@ -110,17 +128,37 @@ func prepareBundle(cmdCtx *cmdcontext.CmdCtx, packCtx PackCtx,
...
+
 	{
 		appsToPack := ""
 		for _, appInfo := range appList {
 			appsToPack += appInfo.Name + " "
 		}
+		if packCtx.CartridgeCompat {
+			if packCtx.Name != "" {
+				// Need to change application name.
+				appsToPack += ", which will be named " + packCtx.Name
```
[the remark this answers is not in the document]
Comment (DerekBum) [reply cli/pack/common.go:144]: You need to use `--name` flag with argument, that differs from the application name

**Review by psergee — APPROVED** [review]
lgtm

**Review by LeonidVas — CHANGES_REQUESTED** [review]
Hi! Thank you for the patch.
Generally ok I only have one nit.

**Review by LeonidVas — APPROVED** [review]

**Pull request merged**